storage/innobase/handler/ha_innodb.cc

```diff
@@ -9489,6 +9490,62 @@ ha_innobase::rnd_next(
 	DBUG_RETURN(error);
 }
 
+#include "../row/row0sel.cc"
+
+int
+ha_innobase::sample_init()
+{
+  int		err;
+
+  /* Store the active index value so that we can restore the original
+  value after a scan */
+
+  if (m_prebuilt->clust_index_was_generated) {
+    err = change_active_index(MAX_KEY);
+  } else {
+    err = change_active_index(m_primary_key);
+  }
+
+  return(err);
+}
+
+int
+ha_innobase::sample_end()
+{
+  return(index_end());
+}
+int
+ha_innobase::sample_next(
+/*=====================*/
+    uchar *buf)
+{
+  mtr_t		mtr;
+  btr_pcur_t*	pcur= m_prebuilt->pcur;
+  rec_t*          rec;
+  rec_offs	offsets_[REC_OFFS_NORMAL_SIZE];
+  rec_offs*	offsets= offsets_;
+  rec_offs_init(offsets_);
+  mtr.start();
+  dict_index_t*	index= innobase_get_index(MAX_KEY);
+  bool res= btr_pcur_open_at_rnd_pos(index, BTR_SEARCH_LEAF, pcur, &mtr);
+  if(!res) {
+    mtr.commit();
+    return HA_ERR_KEY_NOT_FOUND;
+  }
+
+  rec= btr_pcur_get_rec(pcur);
+
+  mem_heap_t*	heap= NULL;
+  offsets= rec_get_offsets(rec, index, offsets, index->n_core_fields, ULINT_UNDEFINED, &heap);
+  ut_ad(offsets != NULL);
+  ut_ad(heap == NULL);
```

Comment:
Good then, we had some argue about that -- whether we should handle `heap` or not. 
@MagHErmit, you should then construct the test to check this out:) And don't throw it out afterwards, we will include it in the test suite.

The plan minimum is to free it correctly, but, Marko, maybe we should cache this heap in `handler` or `prebuilt` to avoid reallocations on each sample_next for wide tables?